pkg/manager/member/pd_failover.go

```diff
@@ -163,50 +164,45 @@ func (f *pdFailover) tryToDeleteAFailureMember(tc *v1alpha1.TidbCluster) error {
-	err = controller.GetPDClient(f.deps.PDControl, tc).DeleteMemberByID(memberID)
-	if err != nil {
-		klog.Errorf("pd failover: failed to delete member: %d, %v", memberID, err)
+	if err := controller.GetPDClient(f.deps.PDControl, tc).DeleteMemberByID(memberID); err != nil {
+		klog.Errorf("pd failover: failed to delete member %d, %v", memberID, err)
 		return err
 	}
-	klog.Infof("pd failover: delete member: %d successfully", memberID)
-	f.deps.Recorder.Eventf(tc, apiv1.EventTypeWarning, "PDMemberDeleted",
-		"%s(%d) deleted from cluster", failurePodName, memberID)
+	klog.Infof("pd failover: delete member %d successfully", memberID)
+	f.deps.Recorder.Eventf(tc, apiv1.EventTypeWarning, "PDMemberDeleted", "failure member %s(%d) deleted from PD cluster", failurePodName, memberID)
 
 	// The order of old PVC deleting and the new Pod creating is not guaranteed by Kubernetes.
 	// If new Pod is created before old PVC deleted, new Pod will reuse old PVC.
-	// So we must try to delete the PVC and Pod of this PD peer over and over,
-	// and let StatefulSet create the new PD peer with the same ordinal, but don't use the tombstone PV
+	// In this case, there is a period during which both new & old Pod is using the same PVC.
+	// So we must try to delete the PVC and Pod of this PD peer over and over again,
+	// and let StatefulSet to create the new PD peer with the same ordinal, but not to use the tombstone PV
 	pod, err := f.deps.PodLister.Pods(ns).Get(failurePodName)
 	if err != nil && !errors.IsNotFound(err) {
 		return fmt.Errorf("tryToDeleteAFailureMember: failed to get pods %s for cluster %s/%s, error: %s", failurePodName, ns, tcName, err)
 	}
 
-	ordinal, err := util.GetOrdinalFromPodName(failurePodName)
-	if err != nil {
-		return err
-	}
-	pvcName := ordinalPVCName(v1alpha1.PDMemberType, controller.PDMemberName(tcName), ordinal)
-	pvc, err := f.deps.PVCLister.PersistentVolumeClaims(ns).Get(pvcName)
-	if err != nil && !errors.IsNotFound(err) {
-		return fmt.Errorf("tryToDeleteAFailureMember: failed to get pvc %s for cluster %s/%s, error: %s", pvcName, ns, tcName, err)
-	}
-
-	if pod != nil && pod.DeletionTimestamp == nil {
-		err := f.deps.PodControl.DeletePod(tc, pod)
-		if err != nil {
-			return err
+	if pod != nil {
+		pvcs, err := util.ResolvePVCFromPod(pod, f.deps.PVCLister)
+		if err != nil && !errors.IsNotFound(err) {
+			return fmt.Errorf("tryToDeleteAFailureMember: failed to get pvcs for pod %s/%s in tc %s/%s, error: %s", ns, pod.Name, ns, tcName, err)
 		}
-	}
-	if pvc != nil && pvc.DeletionTimestamp == nil && pvc.GetUID() == failureMember.PVCUID {
-		err = f.deps.PVCControl.DeletePVC(tc, pvc)
-		if err != nil {
-			klog.Errorf("pd failover: failed to delete pvc: %s/%s, %v", ns, pvcName, err)
-			return err
+		if pod.DeletionTimestamp == nil {
+			if err := f.deps.PodControl.DeletePod(tc, pod); err != nil {
+				return err
+			}
+		}
+		for _, pvc := range pvcs {
+			if pvc.DeletionTimestamp == nil && pvc.GetUID() == failureMember.PVCUID {
```

Comment:
Only one PVC is recorded in the failureMember, what about the other PVCs?

Reply:
failover related FailureMembers/UnjoinedMembers will be fixed in another PR